consumer/shard.go

```diff
@@ -232,7 +236,9 @@ func servePrimary(s *shard) (err error) {
 		defer t.Stop()
 		hintsCh = t.C
 	}
+
 	// Run consumer transactions until an error occurs (such as context.Cancelled).
+	s.publisher = message.NewPublisher(s.ajc, &s.clock)
 	for {
 		var msgCh = make(chan EnvelopeOrError, messageBufferSize)
 
```

Comment:
When a consumer signals they would like to close the transaction via `ErrDeferToNextTransaction`, we nil out the `txn.readCh`.
1. Do we need to explicitly close this channel?
2. Is this the only mechanism needed to resume? Or is there more to the contract around using `ErrDeferToNextTransaction`?

On the Flow side, will this cause any `Task`/`Application` to restart itself when the channel is closed and we subsequently call `StartReadingMessages` again? I'm looking at `Captures` since it's a `MessageProducer`, but this could matter for non-`MessageProducer`s too.

Specifically, `Capture.serverDriverTransactions`, which will attempt to send data to this channel that we're setting to `nil`. I'm not sure how (or how far) this propagates back to the Capture driver, but it seems like we either need to maintain some continuity or deliberately restart things.

Reply:
1) No, we don't want to close the channel. Only senders should ever close a channel in Go, and we're a receiver. In fact, `readCh` isn't a type `chan Message` but rather a type `<-chan Message`, meaning it can only receive and cannot be closed. This particular channel gets closed by the implementation of `StartReadingMessages` that pumps into it, when that function decides it's done for whatever reason and isn't going to send anymore.

2) The effect of this is to not read any more messages _in this transaction_. The `readCh` variable on the *transaction is a separate reference from the `readCh` passed to `runTransactions`. Each `transaction` starts with the latter, and sets it to nil when it's not going to read anymore. `txnInit` initializes the next `transaction` with a non-nil `readCh` and so on.

Reply:
> The effect of this is to not read any more messages in this transaction. The readCh variable on the *transaction is a separate reference from the readCh passed to runTransactions. Each transaction starts with the latter, and sets it to nil when it's not going to read anymore. txnInit initializes the next transaction with a non-nil readCh and so on.

Gotcha! This is what I was missing here. The channel continues to fill with new messages, even as the transaction goes through the motions of committing itself. :100: 

> Only senders should ever close a channel in Go

How does a receiver signal that it's done? Would you open yet another channel to signal that we're hanging up?

Reply:
| How does a receiver signal that it's done?

This is really where `Context` comes in. You cancel a context that the sender is also selecting over.

Naively, just having the receiver go away is a good way to get leaked goroutnines trying to send into a channel that will never again unblock. So you always want another avenue by which the sender discovers that it's work isn't needed anymore. Check out the StartReadingMessages pump in go/shuffle/reader.go in the Flow codebase; it's send is a select over sending a message OR receiving <-ctx.Done().

Also check out `servePrimary` in shard.go. When we want to do a clean restart of the message loop, that's done by having the _sender_ close the channel, and when we've read through its contents and the close, then we start up a new message loop.